From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Wed, 17 Sep 2025 10:14:14 +0900
Subject: [PATCH] Ignore lookups if closed

Project: Pufferfish (https://github.com/pufferfish-gg/Pufferfish)
Author: Kevin Raneri <kevin.raneri@gmail.com>
License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)

Description:
* This patch is based on Pufferfish's "Ignore lookups if closed" patch.

diff --git a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
index 8c5a5043ad8b69d55fa1fc99864767238d435991..92b02f31303ba0728d0dbc07510ac893a9504190 100644
--- a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
@@ -50,6 +50,7 @@ public final class PluginClassLoader extends URLClassLoader implements io.paperm
     private java.util.logging.Logger logger; // Paper - add field
     private io.papermc.paper.plugin.provider.classloader.PluginClassLoaderGroup classLoaderGroup; // Paper
     public io.papermc.paper.plugin.provider.entrypoint.DependencyContext dependencyContext; // Paper
+    private boolean closed = false; // Plazma - Ignore lookups if closed
 
     static {
         ClassLoader.registerAsParallelCapable();
@@ -212,7 +213,7 @@ public final class PluginClassLoader extends URLClassLoader implements io.paperm
         }
         Class<?> result = classes.get(name);
 
-        if (result == null) {
+        if (result == null && !this.closed) { // Plazma - Ignore lookups if closed
             String path = name.replace('.', '/').concat(".class");
             // Add details to zip file errors - help debug classloading
             JarEntry entry;
@@ -268,6 +269,7 @@ public final class PluginClassLoader extends URLClassLoader implements io.paperm
             this.setClass(name, result); // Paper
         }
 
+        if (result == null) throw new ClassNotFoundException(name); // Plazma - Ignore lookups if closed
         return result;
     }
 
@@ -283,6 +285,7 @@ public final class PluginClassLoader extends URLClassLoader implements io.paperm
             super.close();
         } finally {
             jar.close();
+            this.closed = true; // Plazma - Ignore lookups if closed
         }
     }
 
